lib/improver/utilities/statistical_operations.py

```diff
@@ -0,0 +1,256 @@
+        percentile coordinate.
+                e.g.  0th Percentile - Value = 10
+                     10th Percentile - Value = 5
+                     20th Percenitle - Value = 0
+
+        Args:
+            percentiles_cube (iris.cube.Cube):
+                The percentiled field from which probabilities will be obtained
+                using the input cube. This cube should contain a percentiles
+                dimension, with fields of values that correspond to these
+                percentiles. The cube passed to the process method will contain
+                values of the same diagnostic (e.g. height) as this reference
+                cube.
+            output_name (str):
+                The name of the cube being created,
+                e.g.'probability_of_snowfall'.
+        """
+        self.percentile_coordinate = find_percentile_coordinate(
+            percentiles_cube)
+        self.percentiles_cube = percentiles_cube
+
+        if output_name is not None:
+            self.output_name = output_name
+        else:
+            self.output_name = "probability_of_{}".format(
+                percentiles_cube.name())
+
+        # Set inverse_ordering switch
+        percentile_slices = percentiles_cube.slices_over(
+            self.percentile_coordinate)
+        self.inverse_ordering = False
+        first_percentile = percentile_slices.next().data
+        for percentile_values in percentile_slices:
+            last_percentile = percentile_values.data
+        if (first_percentile - last_percentile >= 0).all():
+            self.inverse_ordering = True
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<ProbabilitiesFromPercentiles2D: percentiles_cube: {}, '
+                  'output_name: {}'.format(self.percentiles_cube,
+                                           self.output_name))
+        return result
+
+    def create_probability_cube(self, cube):
+        """
+        Create a 2-dimensional probability cube in which to store the
+        calculated probabilities.
+
+        Args:
+            cube (iris.cube.Cube):
+                Template for the output probability cube.
+        Returns:
+            probability_cube (iris.cube.Cube):
+                A new 2-dimensional probability cube with suitable metadata.
+        """
+        cube_format = next(cube.slices([cube.coord(axis='y'),
+                                        cube.coord(axis='x')]))
+        probabilities = cube_format.copy(data=np.full(cube_format.shape,
+                                                      np.nan, dtype=float))
+        try:
+            probabilities.remove_coord(self.percentile_coordinate)
+        except CoordinateNotFoundError:
+            pass
+
+        probabilities.units = 1
+        probabilities.rename(self.output_name)
+        return probabilities
+
+    def percentile_interpolation(self, threshold_cube, percentiles_cube):
```

Comment:
I think this whole method needs more comments or explanation throughout it. When you first look at it you do have to think carefully about what each step is doing in order to understand it.

Reply:
I've added a truly enormous doc-string explanation. It's rather a simple process, but it takes quite some explaining.